File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import math

def solution(fees, records):
    answer = []
    # 기본 시간, 기본 요금, 단위 시간, 단위 요금
    default_time, default_fee, unit_time, unit_fee = fees

    time_records = {}
    for record in records:
        record = record.split()
        time, car_num, history_type = record
        time = list(map(int, time.split(":")))
        time = time[0] * 60 + time[1]
        if car_num in time_records:
            time_records[car_num].append([time, car_num, history_type])
        else:
            time_records[car_num] = [[time, car_num, history_type]]

    # 입차된 후에 출차된 내역이 없으면 23:59 추가
    for item in time_records:
        if len(time_records[item]) % 2 == 1:
            time_records[item].append([1439, item, "OUT"])
        
    # 요금 계산
    for record in time_records:
        total_time = 0
        for i in range(0, len(time_records[record]) - 1, 2):
            in_record = time_records[record][i]
            out_record = time_records[record][i + 1]
            total_time += out_record[0] - in_record[0]
        total_fee = 0
        if total_time > default_time:
            total_fee = default_fee + math.ceil((total_time - default_time) / unit_time) * unit_fee
        else:
            total_fee = default_fee
        answer.append([record, total_fee])
    
    answer = sorted(answer, key = lambda x : x[0])
    answer = [item[1] for item in answer]

    return answer

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.js
// 시각을 입력받은 분으로 변환하는 함수
const convertMin = (time) => {
    const splitedTime = time.split(":").map((t) => Number(t));
    return splitedTime[0] * 60 + splitedTime[1];
}

function solution(fees, records) {
    const [defaultTime, defaultFee, unitTime, unitFee] = fees;
    const answer = [];
    const inMap = new Map();
    const outMap = new Map();
    const resultMap = new Map();
    
    for (const record of records) {
        const [time, carNum, type] = record.split(" ");
        
        if (type === "IN") {
            // 입차인 경우
            inMap.set(carNum, convertMin(time));
        } else if (type === "OUT") {
            // 출차인 경우  
            const resultTime = convertMin(time) - inMap.get(carNum);
            resultMap.set(carNum, resultMap.has(carNum) ? resultMap.get(carNum) + resultTime : resultTime);
            inMap.delete(carNum)
        }
    }
    
    // 출차 이력이 없는 차량 계산
    if (inMap.size) {
        for (const key of inMap.keys()) {
            const resultTime = convertMin("23:59") - inMap.get(key);
            resultMap.set(key, resultMap.has(key) ? resultMap.get(key) + resultTime : resultTime);
            inMap.delete(key)
        }
    }
    
    // 요금계산
    for (const [key, value] of resultMap.entries()) {
        answer.push([key, value > defaultTime ? defaultFee + (Math.ceil((value - defaultTime) / unitTime) * unitFee)  : defaultFee])
    }
    
    return answer.sort((a, b) => a[0] - b[0]).map((car) => car[1]);
}